nuxeo-services/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoCorsCsrfFilter.java

```diff
@@ -172,6 +230,99 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
+        return GET.equals(method) || HEAD.equals(method) || OPTIONS.equals(method) || TRACE.equals(method);
+    }
+
+    /**
+     * Manages the CSRF token.
+     * <p>
+     * This method may return a response with token fetch information or with an error if needed, in which case it will
+     * return {@code true}.
+     *
+     * @return {@code true} if the caller doesn't need to do more work (a response has been sent)
+     * @since 10.3
+     */
+    protected boolean manageCSRFToken(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        if (!csrfTokenEnabled) {
+            log.debug("No CSRF token check configured");
+            return false; // no check to do
+        }
+
+        String method = request.getMethod();
+        String path = request.getServletPath();
+        if (path == null) {
+            path = "";
+        }
+        String pathInfo = request.getPathInfo();
+        if (pathInfo != null) {
+            path += pathInfo;
+        }
+        String requestToken = request.getHeader(CSRF_TOKEN_HEADER);
+
+        // token fetch request
+        if (GET.equals(method) && path.isEmpty() && CSRF_TOKEN_FETCH.equals(requestToken)) {
+            HttpSession session = request.getSession(); // create if needed
+            String token = (String) session.getAttribute(CSRF_TOKEN_ATTRIBUTE);
+            if (token == null) {
+                token = generateNewToken();
+                session.setAttribute(CSRF_TOKEN_ATTRIBUTE, token);
+            }
+            log.debug("Returning CSRF token fetch");
+            response.setHeader(CSRF_TOKEN_HEADER, token);
+            response.setStatus(SC_OK);
+            return true;
+
+        }
+
+        // do we need to check the token
+        if (isSafeMethod(method)) {
+            log.debug("No CSRF token check on safe method");
+            return false;
+        }
+
+        // is the endpoint specially configured to skip the token check?
+        if (csrfTokenSkipPaths.contains(path)) {
+            log.debug("No CSRF token check on configured endpoint");
+            return false;
+        }
+
+        // check the token
+        HttpSession session = request.getSession(false);
+        String token;
+        if (session == null || (token = (String) session.getAttribute(CSRF_TOKEN_ATTRIBUTE)) == null) {
+            log.debug("Error, no session or no CSRF token in session");
+            String message = "CSRF check failure";
+            log.warn(message + ": invalid token");
+            response.sendError(SC_FORBIDDEN, message);
+            return true;
+        }
+        if (StringUtils.isEmpty(requestToken)) {
+            // allow request parameter to contain the token too
+            requestToken = request.getParameter(CSRF_TOKEN_PARAM);
+        }
+        if (!token.equals(requestToken)) {
+            log.debug("Error, CSRF token does not match");
+            String message = "CSRF check failure";
```

Comment:
Does it worth to make a constant ? At least two occurrences.

Reply:
I don't like doing it for error messages, it obscures things.